paraview/gclfield2vtk/gclfield2vtk.1
.TH GCLFIELD2VTK 1
.SH NAME
gclfield2vtk \- convert internal gclfield objects to vtk format files
.SH SYNOPSIS
.nf
\fBgclfield2vtk\fR db|infile outfile [-i | -g gridname -f fieldname] 
             -odbf outfieldname -r -xml|-binary -pf pffile] -V
.fi
.SH DESCRIPTION
.LP
This program is used to convert Geographical CurviLinear (GCL) field objects
stored in an Antelope database to external formats that can be read by 
the Visualization Tool Kit (VTK).  The program can convert scalar3d, vector3d,
scalar2d, and grid2d (structured grid surfaces such as a DEM) objects.  
Outputs are in "legacy VTK" format or the newer XML format depending on the 
choice of options.  Not all combinations are currently supported.  
(see Bugs and Caveats section).
.LP
Argument 1 is required.  If the \fB-i\fR flag is present this is assumed to
be an input file base name.  Otherwise it is assumed to be a database
named \fBdb\fR.  In the later case it must point to a valid Antelope database
descriptor.  \fBoutfile\fR is the (required) output file name.  Use a simple 
name without an extension for this file.  The program appends an appropriate 
extension depending on the output option.  This is necessary for paraview to
grok the file type in windows, macos, or linux implementations.  The output 
files should normally be readable directly by paraview. 
Currently the extensions are "vtk" for binary or ascii legacy vtk files 
and "vts" for 3D grid output using the -xml option.
.SH OPTIONS
.IP -g 
Override the parameter file grid name to locate the gclfield to be converted.
The current implementation stores GCLfield objects in an antelope database
keyed by a gridname and a fieldname.  This is one key.
This is ignored if the -i flag is used.
.IP -f 
Defines the related key for the field name used to index the desired 
data.  This is ignored if the -i flag is used.
.IP -r
Because visualizations use a Cartesian coordinate system it is always
essential to make sure a set of scene components are in a common reference frame. 
This flag is a switch to turn on this option (default is false). 
When used the program will look for remap parameters in the parameter file 
(see below).
.IP -odbf
This is most commonly used to apply an internal agc operator to 
a field and save the result.  This parameter should be avoided an 
will probably eventually be depricated.
.IP -xml
Use the -xml to have the output written in vtk's xml format.
This is strongly recommended for large 3D structured grid as the resulting
files are platform indepenent files that are much faster to read than
the default ascii file.
.IP -binary
Use the -binary format to have the output written as binary.  This is 
comparable to -xml but the vtk control parameters are written in ascii 
instead of being coded in xml.  This format may be slightly faster to 
read than -xml, but the authors experience is it is not significant if 
true at all.   Generally -xml should be used in preference to -binary 
as the output file is more generic.
.IP -pf
Use pffile.pf as the alternative parameter file to the standard gclfield2vtk.pf.
Note this program does not use the Antelope pf feature to search for pf 
files in a heirarchy of places.  Only the file referenced is read.
.SH FILES
.LP
The default output file is a legacy VTK file described in documentation 
for the Visualization Tool Kit (VTK) library.
Structured grids can optionally be written in the newer XML format also
described in the VTK documentation.
The XML format, in fact, is highly recommended for 3D grids.
.SH PARAMETER FILE
.LP
Two parameters can be viewed as defaults for command line arguments.  
The \fBgridname\fR parameter is the default for -g and
\fBfieldname\fR provides the default for -f.
.LP
The \fBfieldtype\fR defines the type of grid object the program will convert.  Allowed 
options are:  grid2d, scalar2d, scalar3d, and vector3d.  
This parameter must match the type of the data your are attempting to convert
or the program will abort with an error message.
.LP
Several parameters are linked to the -r option:  
\fBorigin_latitude, origin_longitude, radius_origin,\fR and \fBazimuth_y_axis\fR.  
The first three specific the latitude (degrees), longitude (degrees), and radius (km)
of the Cartesian coordinate origin.  The coordinate axes can be rotated by the 
angle specified by \fBazimuth_y_axis\fR.   As the name implies this is the rotation 
in the geographic azimuth sense of the y (north) axis in degrees.  i.e. a positive 
angle rotates the y (north) axis toward the east direction.
.LP
The boolean parameter \fBsave_as_vector_field\fR is a switch for handling vector field data.
When true output is VTK's vector field format.  When false vector fields are output with one 
file per vector component.  If not defined this will default to false.
.LP
The output for 3D vector and scalar fields allow an option for name tags
used by VTK to tag each data tuple.   Two closely linked parameters define
this naming:  (1) \fBscalars_name_tag\fR sets the name used for color 
mapping in paraview and (2) \fBdata_component_names\fR define added name tags
for vector components.   For scalar data best practice is to make the two
names the same.  For vector data choose descriptive names for each 
data component.   Note that the \fBdata_components_names\fR is a list 
defined by the antelope &Tbl{} construct in the parameter file.   Use one
line for each name tag.   Make sure the number of names matches the number of
components in the data you are converting.   The program will not abort if
there is a size mismatch, but you may not get the names set right.  Note for
scalar data the &Tbl construct is still required but in that case there should
only be one name. 
.LP
A \fBscalars_name_tag\fR has a special purpose for grid2d objects and maskedgrid2d objects.
In that case it can servce as a switch to switch between elevation and depth, which have 
a difference in sign.  If \fBscalars_name_tag\fR is Elevation the sign is switched to be
elevation.  Otherwise the depth (in km) is written to the scalars attribute section.
.LP
There are some special parameters used only in certain circumstances.  The user is warned these
may one day be depricated in favor of editing and saving to new loction.  The boolean
parameter \fBremove_mean_x3_slices\fR can be useful for absolute velocity tomography models.  
It removes the average value of each layer (constant x3 slice) before saving the output.
The boolean \fBapply_agc\fR and integer valued parameter \fBagc_operator_length\fR are linked.  
When \fBapply_agc\fR is true an agc operator with a length (in vertical-x3 samples) is applied
before output.  The author has used this frequently for wavefield imaging volumes.
.SH DIAGNOSTICS
.LP
Check stderr for errors.  Most problems will generate a relatively verbose message.
.SH "BUGS AND CAVEATS"
.IP (1)
Not all combinations of output format options are support.  The -xml and 
-binary options will only work on 3D scalar or vector fields.  Either
option will silently be ignored for any 2D object which currently only
use ascii output to vtk files..
.IP (2)
For version 3.8 of paraview on 64 bit platforms a bug seems to exist in libc for reading ascii data created by this program.  The VTK reader for ascii files 
will abort when trying to read numbers with exponents that define tiny
numbers that overflow the exponent field on floats.  (e.g. a number like
1.6698897e-258 will cause the VTK reader for legacy VTK files to throw
an exception and cause paraview, which uses this routine, to abort).  
For this reason use only the -xml or -binary option on 64 bit platforms. 
Later versions of paraview may have fixed this bug.
.IP (3)
The agc and remove mean parameters really do not belong in this program.  There is a good chance these
will be replaced with a more elaborate field editor or implemented as a python module in paraview in
the future.
.SH AUTHOR
.nf
Gary L. Pavlis
Department of Geological Sciences
Indiana University 
1001 East 10th Street
Bloomington, IN 47405
pavlis@indiana.edu
.fi


paraview/gclfield2vtk/gclfield2vtk.cc
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <list>
#include <vector>
#include <stdio.h>
#include <float.h>
#include "perf.h"
#include "stock.h"
#include "pf.h"
#include "seispp.h"
#include "Metadata.h"
#include "gclgrid.h"
#include "GCLMasked.h"
#include "agc.h"
#include "vtk_output.h"

using namespace SEISPP;

int vtk_output_GCLgrid(GCLgrid& g, string ofile,string scalars_tag="Elevation");
/* Removes mean for constant x3 slices.  Important for
proper display of tomography models showing absolute velocities */
void remove_mean_x3(GCLscalarfield3d& f)
{
	double mean;
	int i,j,k;
	double nval=static_cast<double>((f.n1)*(f.n2));
	cout << "Mean removal values"<<endl
		<< "grid-index     mean"<<endl;
	for(k=0;k<f.n3;++k)
	{
		mean=0.0;
		for(i=0;i<f.n1;++i)
			for(j=0;j<f.n2;++j)
				mean += f.val[i][j][k];
		mean /= nval;
		cout << k << "      "<<mean<<endl;
		for(i=0;i<f.n1;++i)
			for(j=0;j<f.n2;++j)
				f.val[i][j][k] = f.val[i][j][k]-mean;
	}
}
/* Applies an agc operator with window length iwagc to data along the x3 direction
then normalizes a vector constructed from each x3 grid line to have rms=1 (L2 norm unity).
*/
void agc_scalar_field(GCLscalarfield3d& g, int iwagc)
{
	vector<double> x3vals;
	double nrmx3v;
	int i,j,k;
	for(i=0;i<g.n1;++i)
		for(j=0;j<g.n2;++j)
		{
			x3vals.clear();
			for(k=0;k<g.n3;++k) x3vals.push_back(g.val[i][j][k]);
			x3vals=agc<double>(x3vals,iwagc);
			nrmx3v=dnrm2(g.n3,&(x3vals[0]),1);
			/* do nothing if nrm is 0 */
			if(nrmx3v<DBL_EPSILON) nrmx3v=1.0;
			for(k=0;k<g.n3;++k) g.val[i][j][k]=x3vals[k]/nrmx3v;
		}
}
vector<string> list_to_vector(list<string> l)
{
    list<string>::iterator lptr;
    vector<string> result;
    result.reserve(l.size());
    for(lptr=l.begin();lptr!=l.end();++lptr)
        result.push_back(*lptr);
    return(result);
}
void usage()
{
	cerr << "gclfield2vtk db|file outfile [-i -g gridname -f fieldname -r "
		<< "-odbf outfieldname -xml -binary -pf pffile] -V" << endl;
	exit(-1);
}
bool SEISPP::SEISPP_verbose(true);
int main(int argc, char **argv)
{
	int i,j;

        bool dbmode(true);  // default is db input mode
        ios::sync_with_stdio();
	if(argc<3) usage();
	string dbname(argv[1]);
        string infile(argv[1]);  // redundant but easier to do this way
	string outfile(argv[2]);
	string argstr;
	string pffile("gclfield2vtk.pf");
	const string nodef("NOT DEFINED");
	string gridname(nodef);
	string fieldname(nodef);
	bool saveagcfield(false);
	string outfieldname;
	bool remap(false);
	bool xmloutput(false);
	bool binaryout(false);
	for(i=3;i<argc;++i)
	{
		argstr=string(argv[i]);
		if(argstr=="-pf")
		{
			++i;
			if(i>=argc)usage();
			pffile=string(argv[i]);
		}
                else if(argstr=="-i")
                    dbmode=false;
		else if(argstr=="-V")
		{
                    usage();
		}
		else if(argstr=="-g")
		{
			++i;
			if(i>=argc)usage();
			gridname=string(argv[i]);
		}
		else if(argstr=="-f")
		{
			++i;
			if(i>=argc)usage();
			fieldname=string(argv[i]);
		}
		else if(argstr=="-r")
		{
			remap=true;
		}
		else if(argstr=="-odbf")
		{
			++i;
			if(i>=argc)usage();
			saveagcfield=true;
			outfieldname=string(argv[i]);
		}
		else if(argstr=="-xml")
		{
			xmloutput=true;
                        binaryout=false;
		}
		else if(argstr=="-binary")
		{
			binaryout=true;
                        xmloutput=false;
		}
		else
		{
			usage();
		}
	}
	try {
                
        	PfStyleMetadata control=pfread(pffile);
                DatascopeHandle dbh;
                if(dbmode)
                    dbh=DatascopeHandle(dbname,true);

		if(gridname==nodef)
			gridname=control.get_string("gridname");
		if(fieldname==nodef)
			fieldname=control.get_string("fieldname");
		string fieldtype=control.get_string("fieldtype");
		cout << "Converting field = "<<fieldname
			<< " associated with gridname="<<gridname<<endl;
		cout << "Assuming grid type is "<<fieldtype<<endl;
                int nv_expected(5);  // defaulted for pwmig
                if(fieldtype=="vector3d")
                {
                    nv_expected=control.get_int("nv_expected");
                }
                string scalars_tag=control.get_string("scalars_name_tag");
                list<string> complist=control.get_tbl(string("data_component_names"));
                vector<string> component_names=list_to_vector(complist);
		/* Slightly odd logic here, but this allows remap off
		to be the default.  pf switch is ignored this way if
		the -r flag was used */
		if(!remap) remap=control.get_bool("remap_grid");
		BasicGCLgrid *rgptr;
                rgptr=NULL;
		if(remap)
		{
			cout << "Remapping enabled"<<endl;
                        /* We create a small temporary GCLgrid to allow
                           use of remap_grid procedure.   */
                        double lat0,lon0,r0,azm;
                        lat0=control.get_double("latitude_origin");
                        lon0=control.get_double("longitude_origin");
                        r0=control.get_double("radius_origin");
                        azm=control.get_double("azimuth_y_axis");
                        cout << "Origin lat,lon,r="
                            <<lat0<<", "<<lon0<<", "<<r0<<endl;
                        cout << "Coordinate system y axis rotation="
                            <<azm<<endl;
                        lat0=rad(lat0);
                        lon0=rad(lon0);
                        azm=rad(azm);
                        GCLgrid *gtmp=new GCLgrid(2,2,string("reference"),
                                lat0,lon0,r0,azm,
                                1.0,1.0,0,0);
                        rgptr=dynamic_cast<BasicGCLgrid*>(gtmp);

		}
			
		bool SaveAsVectorField;
		int VectorComponent;
		bool rmeanx3=control.get_bool("remove_mean_x3_slices");
                if(rmeanx3)
                    cout << "remove mean for each slice enabled"<<endl;
		bool apply_agc=control.get_bool("apply_agc");
		int iwagc(0);
		if(apply_agc)
		{
			iwagc=control.get_int("agc_operator_length");
                        cout << "Applying agc operator with length="<<iwagc
                            << " depth intervals"<<endl;
		}
                else
                    cout << "Date being written with true amplitude"<<endl;
		if(saveagcfield  && !apply_agc)
		{
			cerr << "Illegal option combination"<<endl
			  << "-odbf argument can only be used if apply_agc"
			  << " pf parameter is set true"<<endl;
			exit(-1);
		}
		string fielddir;
		if(saveagcfield)
			fielddir=control.get_string("field_directory");
		if(fieldtype=="scalar3d") 
		{
                        GCLscalarfield3d field;
                        if(dbmode)
			    field=GCLscalarfield3d(dbh,gridname,fieldname);
                        else
                            field=GCLscalarfield3d(infile);
			if(remap)
			{
				// Used to make this optional.  force
				//if(field!=(*rgptr))
				remap_grid(dynamic_cast<GCLgrid3d&>(field),
						*rgptr);
			}
			if(rmeanx3) remove_mean_x3(field);
			if(apply_agc) agc_scalar_field(field,iwagc);
                        if(xmloutput)
                            outfile=outfile+".vts";
                        else
                            outfile=outfile+".vtk";
			output_gcl3d_to_vtksg<GCLscalarfield3d&>(field,outfile,
                               scalars_tag,component_names,xmloutput,binaryout);
			if(saveagcfield) 
				field.save(dbh,string(""),fielddir,
				  outfieldname,outfieldname);
		}
		else if(fieldtype=="vector3d")
		{
			SaveAsVectorField=control.get_bool("save_as_vector_field");
			if(rmeanx3)
			{
				cerr << "remove_mean_x3_slices set true:  "
					<< "ignored for vector field="
					<< fieldname<<endl;
			}
                        GCLvectorfield3d vfield;
                        if(dbmode)
			    vfield=GCLvectorfield3d(dbh,gridname,
                                    fieldname,nv_expected);
                        else
                            vfield=GCLvectorfield3d(infile);
			if(remap)
			{
				//if(vfield!=(*rgptr))
					remap_grid(dynamic_cast<GCLgrid3d&>(vfield),
						*rgptr);
			}
                        if(SaveAsVectorField)
                        {
                            if(xmloutput)
                                outfile=outfile+".vts";
                            else
                                outfile=outfile+".vtk";
                            output_gcl3d_to_vtksg<GCLvectorfield3d&>(vfield,outfile,
                                    scalars_tag,component_names,
                                    xmloutput,binaryout);
                        }
                        else
                        {
			    for(i=0;i<vfield.nv;++i)
			    {
			        GCLscalarfield3d *sfptr;
                                vector<string> thiscomponent;
				sfptr = extract_component(vfield,i);
				if(apply_agc) agc_scalar_field(*sfptr,iwagc);
				stringstream ss;
				ss << outfile <<"_"<<i;
				if(xmloutput) 
					ss<<".vts";
				else
					ss<<".vtk";
                                thiscomponent.clear();
                                thiscomponent.push_back(component_names[i]);
				output_gcl3d_to_vtksg<GCLscalarfield3d&>(*sfptr,ss.str(),
                                        scalars_tag,thiscomponent,
                                        xmloutput,binaryout);
				/*This is not ideal, but will do this now
				for expedience.  This creates a series of 
				scalar fields when saveagcfield is enabled
				that have a naming convention similar to
				the vtk files.  These could easily overflow
				but I don't test them here. beware */
				if(saveagcfield) 
				{
					stringstream ssof;
					ssof << outfieldname<<"_"<<i;
					string ofld=ssof.str();
					sfptr->save(dbh,string(""),
					  fielddir,ofld,ofld);
				}
				delete sfptr;
                            }
			}
		}
		else if(fieldtype=="grid2d")
		{
			int npoly;
                        GCLgrid g;
                        if(dbmode)
			    g=GCLgrid(dbh,gridname);
                        else
                            g=GCLgrid(infile);
			if(remap)
			{
				//if(g!=(*rgptr))
				remap_grid(g,*rgptr);
			}
                        outfile=outfile+".vtk";
			npoly=vtk_output_GCLgrid(g,outfile,scalars_tag);
			cout << "Wrote "<<npoly<<" polygons to output file"<<endl;
			if(apply_agc)cerr <<"WARNING: apply_agc was set true\n"
					<<"This is ignored for grids and 2d fields"<<endl;
		}
		else if(fieldtype=="scalar2d")
		{
			int npoly;
                        GCLscalarfield field;
			if(dbmode)
                            field=GCLscalarfield(dbh,gridname,fieldname);
                        else 
                            field=GCLscalarfield(infile);
			if(remap)
			{
			    remap_grid(dynamic_cast<GCLgrid&>(field),
						*rgptr);
			}
                        outfile=outfile+".vtk";
			npoly=vtk_output_GCLgrid(field,outfile,scalars_tag);
			cout << "Wrote "<<npoly<<" polygons to output file"<<endl;
			if(apply_agc)cerr <<"WARNING: apply_agc was set true\n"
					<<"This is ignored for grids and 2d fields"<<endl;
		}	
		else if( (fieldtype=="MaskedScalar2d")
                        || (fieldtype=="MaskedGrid2d"))
		{
			int npoly;
                        GCLMaskedScalarField *field;
			if(dbmode)
                        {
                            cerr << "Not current support for db input for masked grid objects"
                                <<endl
                                << "Probably a parameter file inconsistency"<<endl;
                            exit(-1);
                        }
                        else if(fieldtype=="MaskedScalar2d")
                            field=new GCLMaskedScalarField(infile);
                        else
                        {
                            /* Logic assumes can only land here if fieldtype is 
                               masked 2d grid.  In that case we read the grid and
                               add depth/elevation as an attribute */
                            GCLMaskedGrid gtmp(infile);
                            field=new GCLMaskedScalarField(gtmp);
                            if(scalars_tag=="Elevation")
                            {
                                for(i=0;i<gtmp.n1;++i)
                                    for(j=0;j<gtmp.n2;++j)
                                    {
                                        if(field->point_is_valid(i,j))
                                            field->val[i][j]=(-gtmp.depth(i,j));
                                        else
                                            field->val[i][j]=0.0;
                                    }
                            }
                            else
                            {
                                for(i=0;i<gtmp.n1;++i)
                                    for(j=0;j<gtmp.n2;++j)
                                    {
                                        if(field->point_is_valid(i,j))
                                            field->val[i][j]=gtmp.depth(i,j);
                                        else
                                            field->val[i][j]=0.0;
                                    }
                            }
                        }
			if(remap)
			{
			    remap_grid(dynamic_cast<GCLgrid&>(*field),
						*rgptr);
			}
                        outfile=outfile+".vtk";
			npoly=vtk_output_GCLgrid(*field,outfile,scalars_tag);
			cout << "Wrote "<<npoly<<" polygons to output file"<<endl;
			if(apply_agc)cerr <<"WARNING: apply_agc was set true\n"
					<<"This is ignored for grids and 2d fields"<<endl;
		}	
		else
		{
			cerr << "Unsupported fieldtype = "<<fieldtype<<endl
				<< "Exiting with no output\n" << endl;
		}

	}
	catch (MetadataGetError& mderr)
	{
		mderr.log_error();
		exit(-1);
	}
        catch(GCLgridError& gerr)
        {
            cerr << "GCLgridError thrown"<<endl<<gerr.what();
        }
        catch(std::exception& sexcp)
        {
            cerr << sexcp.what();
        }
        catch(SeisppError& serr)
        {
            serr.log_error();
        }
	catch (...)
	{
	    cerr << "Something threw an unhandled exception"<<endl;
	}
}
